pkg/plugins/runtime/k8s/controllers/meshservice_controller.go

```diff
@@ -0,0 +1,202 @@
+type MeshServiceReconciler struct {
+	kube_client.Client
+	Log    logr.Logger
+	Scheme *kube_runtime.Scheme
+}
+
+func (r *MeshServiceReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("service", req.NamespacedName)
+
+	namespace := &kube_core.Namespace{}
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Name: req.Namespace}, namespace); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to get Namespace for Service")
+	}
+	injectedLabel, _, err := metadata.Annotations(namespace.Labels).GetEnabled(metadata.KumaSidecarInjectionAnnotation)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrapf(err, "unable to check sidecar injection label on namespace %s", namespace.Name)
+	}
+
+	svc := &kube_core.Service{}
+	if err := r.Get(ctx, req.NamespacedName, svc); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return kube_ctrl.Result{}, nil
+		}
+		return kube_ctrl.Result{}, errors.Wrapf(err, "unable to fetch Service %s", req.NamespacedName.Name)
+	}
+
+	_, ok := svc.GetLabels()[mesh_proto.MeshTag]
+	if !ok && !injectedLabel {
+		log.V(1).Info("service is not considered to be service in a mesh")
+		if err := r.deleteIfExist(ctx, req.NamespacedName); err != nil {
+			return kube_ctrl.Result{}, err
+		}
+		return kube_ctrl.Result{}, nil
+	}
+
+	mesh := util.MeshOfByLabelOrAnnotation(log, svc, namespace)
+
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Name: mesh}, &v1alpha1.Mesh{}); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			log.V(1).Info("mesh not found")
+			if err := r.deleteIfExist(ctx, req.NamespacedName); err != nil {
+				return kube_ctrl.Result{}, err
+			}
+			return kube_ctrl.Result{}, nil
+		}
+		return kube_ctrl.Result{}, err
+	}
+
+	ms := &meshservice_k8s.MeshService{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      svc.GetName(),
+			Namespace: svc.GetNamespace(),
+		},
+	}
+	if err := kube_controllerutil.SetOwnerReference(svc, ms, r.Scheme); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "could not set owner reference")
+	}
+
+	operationResult, err := kube_controllerutil.CreateOrUpdate(ctx, r.Client, ms, func() error {
+		ms.ObjectMeta.Labels = maps.Clone(svc.GetLabels())
+		if ms.ObjectMeta.Labels == nil {
+			ms.ObjectMeta.Labels = map[string]string{}
+		}
+		ms.ObjectMeta.Labels[mesh_proto.MeshTag] = mesh
+		if ms.Spec == nil {
+			ms.Spec = &meshservice_api.MeshService{}
+		}
+
+		ms.Spec.Selector = meshservice_api.Selector{
+			DataplaneTags: svc.Spec.Selector,
+		}
+
+		ms.Spec.Ports = []meshservice_api.Port{}
+		for _, port := range svc.Spec.Ports {
+			ms.Spec.Ports = append(ms.Spec.Ports, meshservice_api.Port{
+				Port:       uint32(port.Port),
+				TargetPort: uint32(port.TargetPort.IntVal), // todo(jakubdyszkiewicz): what to do about named ports?
+				Protocol:   core_mesh.Protocol(pointer.DerefOr(port.AppProtocol, "tcp")),
+			})
+		}
+
+		if svc.Spec.ClusterIP != "" {
```

Comment:
What about stateful set?

Reply:
good point. Ignoring for now. Will add handling after Kube UX MADR is merged.